### ISSUE TITLE
Prevent losing tracker.json custom modifications

### ISSUE THREAD

**Issue by MattEqualsCoder** [Vivelin/SMZ3Randomizer#60]
Possibly create a separate file that would allow for overriding tracker.json to avoid losing personal changes.

**Comment by Vivelin** [comment]
If you're creating a separate file (and then somehow merge that with the built-in one), put it in the Documents folder or something so it's easier to find for most people.

**Comment by Vivelin** [comment]
For upcoming version I'll probably just allow for additions only.

It won't really allow for people to change the weight of some responses nor will it be possible to remove lines but keep others, but we could always make another issue afterwards if that's wanted.

**Comment by Vivelin** [comment]
Okay, this is considerably more difficult than I anticipated.

- There's no unique way to identify items right now. Item type should work, but isn't ideal for non-items like deaths, bullshit, etc.
- Do we manually decide for each type what properties are mergeable and which aren't? Just names and phrases? Everything? 
- How do we let someone adjust the weights on specific phrases?

System.Text.Json doesn't support any way of merging, either. Newtonsoft.Json should, I believe, but then we'd have no control over it whatsoever.

**Comment by MattEqualsCoder** [comment]
I've been thinking about this when I started working on some of the YAML changes. I wonder if a change to how we store tracker related text data is in order in general. Perhaps we could keep some of the complex structure in the JSON and not have people update it, but pull out all of the actual names/responses that people may want to modify into a new YAML based file? We could then come up with a way to link the JSON details with the YAML ones.

I think if we do this, it may solve several of these problems in one swoop.

**Comment by MattEqualsCoder** [comment]
So, in regards to this item, I've got a few thoughts.

I think it might be a good idea to split apart the configs a bit into more manageable chunks for people to modify. Instead of one giant tracker.yml file, perhaps have items.yml, responses.yml, and requests.yml for example.

In regards to the merging, I think maybe additions in fine, and we can actually use this to accomplish different "profiles" for tracker. We could potentially have a bare bones "boring business tracker" configs as a baseline without in jokes and confusing names that have very low weights to everything. Then, there can be profiles which will have new yml files, and basically anything there will likely have a much higher weight than the one in the in the base configs. That way, if someone were to make their own profile, then if we add something new, the baseline one will still be there.

**Comment by MattEqualsCoder** [comment]
Technically that could still be doable, though. It could just make a new custom profile or something. Technically could even allow you to pick which profile(s) you want to use for for different files, but that may be getting to be unnecessary.

**Comment by MattEqualsCoder** [comment]
Here's a mockup of kind of what I was thinking as far as the structure goes.
[configs.zip](https://github.com/Vivelin/SMZ3Randomizer/files/9238680/configs.zip)

In the end, I think I kinda changed my mind. I think rather than adding to the names in the default configs, it would probably be best to pull in the names, hints, etc. from the custom profile and use the defaults for things like the internal item id, map positions, etc. The same would be said of icons. I think this balances giving people control and avoiding potential issues from people messing up as it would fall back to the default if something is missing.

In the end, this may not offer every single piece of functionality that may be desired, such as being able to combine the list of names for a specific item. But, we could potentially allow you to stack multiple profiles. That way if you wanted most of the BCU pack but just override a few items, you would only need to specify the specific items in your custom profile yaml file and all that aren't in there to fall back to the BCU profile. But if we made modifications to the BCU file for that specific item, the person would need to manually pull in those changes. Kinda annoying, but it prevents people from having to put in weights of like 100 if they never want to hear stuff from the other configs.

Some notes:

- Location keys are kinda tricky, but I figure RegionName.RoomName.LocationName is probably best. Which LocationName? Good question. Probably best to just pick whichever one would make the most sense. Hopefully having the Region and Room names (if applicable) would help differentiate some of the locations.
- For the requests.yml file, I was thinking to combine the lists between profiles. But, this brings up the question of duplicate requests in both files. Might be something that will need to be factored in.
- For the UI config example, I'm still thinking of having combined items like the bow + silvers and flute + duck being sort of hard coded to be combined into a single spot in the UI, though the position would still be in the config. I just don't see the need to go to adding the ability to customize the menus into the config since I doubt anyone would actually use that. If people want it, we could add it in later.
- Admittedly the SchrodingersString stuff might be a bit complex for people because the indentation changes based on whether you include a weight or not. I wonder if a class for config strings would be best that would have the text and weight prefixed in the yaml would be better like this. These could be converted to SchrodingersStrings on load. It becomes more verbose, but I think it becomes much more readable. Here's an example of what I mean:

SchrodingersString Version:
```
---
Name:
- ExampleName1
- - ExampleName2
  - 0.1
Plural:
- ExampleName3
Article: ''
InternalItemType: 0
Multiple: true
```

My Suggestion:
```
Name:
- Text: ExampleName1
  Weight: 1
- Text: ExampleName2
  Weight: 0.1
Plural:
- Text: ExampleName3
  Weight: 1
Article: ''
InternalItemType: 0
Multiple: true
```

**Comment by MattEqualsCoder** [comment]
Actually in regards to the last bullet, I just noticed there's a custom converter for SchrodingersString. I think the class would by default probably serialize to what I'm suggesting.

**Comment by Vivelin** [comment]
I think we should look at what others have done already. I think it's just been to add new lines in addition, not necessarily to remove old lines. We might not need to take this into account at all if it's not relevant right now.

- For locations, I think we should just take the ID. We're already using that and people would have to look up the _exact_ value anyway. It's less error prone with numbers, I think. 
- Since we'd stack config files, I'd say the most recent one overrides the previous responses.
- On menus and stacking items, I agree. I don't see anyone adding custom items that use this system, and we could always add that later if the need arises.
- And yes, for YAML syntax I agree that the custom converter is pointless. We could still leave out the weights, anyway.

**Comment by MattEqualsCoder** [comment]
Fair. Combining and stacking could allow for us to have multiple packages:

- Default (can't be disabled) - All the boring basic responses it'll fall back to
- Sassy - Adds a lot of the personality responses and silly item names
- BCU - In jokes more specific to betus. Might also limit the NSFW responses to here as well.

This would offer maximum flexibility.

In regards to the locations, the only reason I'd still maybe suggest my approach is I think it might be more easily readable at a quick glance, especially in custom profiles where all you may be specifying are things like responses when you pick up stuff.

**Comment by Vivelin** [comment]
I'd argue defaults should be present in the code itself, where possible. We already do that for a lot, although some things like item names are more difficult, I suppose. We'd just need to be more consistent about it.

Separating sassy responses and BCU-specific responses is a good idea. 

If it's just for readability, YAML supports comments. We could add a more readable name as comments in some sort of empty "template" config.

**Comment by MattEqualsCoder** [comment]
Default items could be created in a constructor or static method in some sort of "ItemsConfig" class. Locations could either be the same or created in the region files themselves when the locations are created. I'm mixed on those.

**Comment by MattEqualsCoder** [comment]
[items.zip](https://github.com/Vivelin/SMZ3Randomizer/files/9247948/items.zip)

Attached is a first pass of doing some conversions. Currently only the following fields end up being in config files:

- Key
- Name
- Article (possibly could remove?)
- Plural
- When Tracked
- Stages

A lot of the other fields can be there, but I think for the most part if we have hints, if multiple, and some other fields specified in the C# code, the configs can stay pretty clean when adding extra sass and whatnot. The C# file is admittedly gross, but if we want the defaults to come from the code, there's not much of an easier way. I fe

**Comment by MattEqualsCoder** [comment]
Oh a couple random things to note. For right now, I'm using just the first name the C# code, but some work will need to be done to go through the list and see if there are other names that need to be pulled in for the defaults. Some examples would be "morph ball" and "morphing ball".

I also pulled in the maps as they are important to keysanity. Still leaving out the compasses since they are worthless AFAIK.

**Comment by Vivelin** [comment]
Instead of "Key", I would call it "Item", at least in the YAML itself. I imagine there's an attribute that controls that. I think Key (or ID, for that matter) might be too technical and confusing for normal users since we have small keys and big keys too.

Otherwise, it looks more or less like I had in mind for now.

**Comment by MattEqualsCoder** [comment]
To clarify, I more meant the article property would probably be in the defaults in the C# and not really in any configs as I don't think they would be commonly overwritten. People will be able to if they want, but I don't think it'll be one we need to have in the sassy/BCU configs.

**Comment by Vivelin** [comment]
Oh yeah, that's fine.

**Comment by MattEqualsCoder** [comment]
[configs.zip](https://github.com/Vivelin/SMZ3Randomizer/files/9274308/configs.zip)

Alright, I've finally got all of the current data reconstructed into the C# and starting yaml files. I still need to actually figure out what text will be in C# vs sassy profile vs BCU profile, but this should give you an idea of the structure. I also probably need to generate example files.

Few notes/questions

- I don't have maps.json currently, but I honestly probably won't be doing much with it as I don't think anyone would be modifying that.
- Currently I still have Row/Column in the configs, but those will be removed once this is finished and I start working on the tracker item window changes
- Just out of curiosity, is there a reason why WhenTrackingJunk, WhenTrackingProgression, etc. are Lists of SchrodingersStrings instead of just SchrodingersString?
- I've split out the configs into separate files to be easier for people to edit. Do you want me to still have them lumped under the TrackerConfig and LocationConfig like they currently are, or would you want me to have each of them accessible via the TrackerConfigProvider?

**Comment by Vivelin** [comment]
The hard-coded location stuff could probably be integrated into the individual regions, right? Unless you think it's better to keep it separated like this. If you do, I'd prefer the type names to just be types instead.

Also, I would probably make the hints configurable.

> * I don't have maps.json currently, but I honestly probably won't be doing much with it as I don't think anyone would be modifying that.

That's fine. The map isn't going to change much so it's fine if this is handled internally.

> * Just out of curiosity, is there a reason why WhenTrackingJunk, WhenTrackingProgression, etc. are Lists of SchrodingersStrings instead of just SchrodingersString?

Nope. Probably just a brain fart.

> * I've split out the configs into separate files to be easier for people to edit. Do you want me to still have them lumped under the TrackerConfig and LocationConfig like they currently are, or would you want me to have each of them accessible via the TrackerConfigProvider?

I don't really have any strong opinions on this, but it's probably better to have them separated so it stays 1:1.

**Comment by MattEqualsCoder** [comment]
Well, currently the classes are all extensions of collections and I think it's best probably to be consistent with the files as well.

What do you mean you'd prefer the type names to be types? Just rename the property?

**Comment by Vivelin** [comment]
I mean instead of writing

```csharp
TypeName = "Randomizer.SMZ3.Regions.Zelda.ThievesTown",
```

you should do

```csharp
Type = typeof(Randomizer.SMZ3.Regions.Zelda.ThievesTown)
```

Since it's in C# now, we don't have to use strings and refer to the types directly. Or we just put it on the regions & rooms themselves.
